pySDC/implementations/problem_classes/generic_spectral.py

```diff
@@ -0,0 +1,357 @@
+        for line, equation in LHS.items():
+            self.spectral.add_equation_lhs(operator, line, equation)
+        return self.spectral.convert_operator_matrix_to_operator(operator)
+
+    def setup_L(self, LHS):
+        """
+        Setup the left hand side of the linear operator L and store it in ``self.L``.
+
+        The argument is meant to be a dictionary with the line you want to write the equation in as the key and the relationship between components as another dictionary. For instance, you can add an algebraic condition capturing a first derivative relationship between u and ux as follows:
+
+        ```
+        Dx = self.get_self.get_differentiation_matrix(axes=(0,))
+        I = self.get_Id()
+        LHS = {'ux': {'u': Dx, 'ux': -I}}
+        self.setup_L(LHS)
+        ```
+
+        If you put zero as right hand side for the solver in the line for ux, ux will contain the x-derivative of u afterwards.
+
+        Args:
+            LHS (dict): Dictionary containing the equations.
+        """
+        self.L = self._setup_operator(LHS)
+
+    def setup_M(self, LHS):
+        '''
+        Setup mass matrix, see documentation of ``GenericSpectralLinear.setup_L``.
+        '''
+        self.diff_index = list(LHS.keys())
+        self.diff_mask = [me in self.diff_index for me in self.components]
+        self.M = self._setup_operator(LHS)
+
+    def setup_preconditioner(self, Dirichlet_recombination=True, left_preconditioner=True):
+        """
+        Get left and right precondioners. A right preconditioner of D2T will result in Dirichlet recombination. 10/10 would recommend!
+
+        Args:
+            Dirichlet_recombination (bool): Basis conversion for right precondioner
+            left_preconditioner (bool): If True, it will interleave the variables and reverse the Kronecker product
+        """
+        sp = self.spectral.sparse_lib
+        N = np.prod(self.init[0][1:])
+
+        Id = sp.eye(N)
+        Pl_lhs = {comp: {comp: Id} for comp in self.components}
+        self.Pl = self._setup_operator(Pl_lhs)
+
+        if left_preconditioner:
+            # reverse Kronecker product
+
+            if self.spectral.useGPU:
+                R = self.Pl.get().tolil() * 0
+            else:
+                R = self.Pl.tolil() * 0
+
+            for j in range(self.ncomponents):
+                for i in range(N):
+                    R[i * self.ncomponents + j, j * N + i] = 1.0
+
+            self.Pl = self.spectral.sparse_lib.csc_matrix(R)
+
+        if Dirichlet_recombination and type(self.axes[-1]).__name__ in ['ChebychevHelper, Ultraspherical']:
+            _Pr = self.spectral.get_Dirichlet_recombination_matrix(axis=-1)
+        else:
+            _Pr = Id
+
+        Pr_lhs = {comp: {comp: _Pr} for comp in self.components}
+        self.Pr = self._setup_operator(Pr_lhs) @ self.Pl.T
+
+    def solve_system(self, rhs, dt, u0=None, *args, skip_itransform=False, **kwargs):
```

Comment:
You could probably write the equation the implicit Euler update with a M matrix using this method ... would make things more cleared than the not your have trying to describe it with words ...